crates/matrix-sdk-crypto/src/identities/user.rs

```diff
@@ -940,17 +961,42 @@ impl OwnUserIdentityData {
 
     /// Mark our identity as verified.
     pub fn mark_as_verified(&self) {
-        self.verified.store(true, Ordering::SeqCst)
+        *self.verified.write().unwrap() = OwnUserIdentityVerifiedState::Verified;
     }
 
-    #[cfg(test)]
-    pub fn mark_as_unverified(&self) {
-        self.verified.store(false, Ordering::SeqCst)
+    /// Mark our identity as unverified.
+    pub(crate) fn mark_as_unverified(&self) {
+        let mut guard = self.verified.write().unwrap();
+        if *guard == OwnUserIdentityVerifiedState::Verified {
+            *guard = OwnUserIdentityVerifiedState::PreviouslyVerifiedButNoLonger;
+        }
     }
 
     /// Check if our identity is verified.
     pub fn is_verified(&self) -> bool {
-        self.verified.load(Ordering::SeqCst)
+        *self.verified.read().unwrap() == OwnUserIdentityVerifiedState::Verified
+    }
+
+    /// True if we verified our own identity at some point in the past.
+    ///
+    /// To reset this latch back to `false`, one must call
+    /// [`OwnUserIdentityData::withdraw_verification()`].
+    pub fn was_previously_verified(&self) -> bool {
+        let verified_state = (*self.verified.read().unwrap()).clone();
```

Comment:
Do we need to clone here?

Reply:
The problem is that the stored verification state is behind a mutex, so we can't just take a reference to it without holding a separate lock guard. So I think our options are
 * clone, like this
 * use a separate variable to hold a lock guard while we do the comparison
 * `matches!`, as you suggest below

`OwnUserIdentityData` is a unit enum, so cloning it is just copying the enum discriminator -- in other words, loading it into a register (which we'd have to do anyway).

In terms of readability, I'm not sure that any of the others are any better. YMMV?

Reply:
fair enough. will make it so.